### PR TITLE
Constraint checks pt 3

### PR DESCRIPTION
**Pull request by apanda**
Run constraint checks before resuming by default. Most failures are
recorded on the log, fatal failures cause resume to fail.

### CODE REVIEW

File: core/worker.cc [hunk 1]
```diff
@@ -105,7 +105,13 @@ static void resume_worker(int wid) {
...
 }
 
-void resume_all_workers() {
+
+/*!
+ * Attach orphan TCs to workers. Note this does not ensure optimal placement.
+ * This method can only be called when all workers are paused.
+ */
+void attach_orphans() {
+  assert(!is_any_worker_running());
```
Comment (melvinw) [inline core/worker.cc]: This should be `CHECK`. We compile with DNEDBUG, so this won't have the intended behavior unless you do a debug build

File: core/modules/sink.h [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
 class Sink final : public Module {
  public:
+  Sink() { max_allowed_workers_ = Worker::kMaxWorkers; }
```
Comment (barath) [inline core/modules/sink.h]: Are there other modules this is true for?  I.e. other modules that are inherently thread safe?

Also, what is the right annotation for the various Queue* modules?

Comment (apanda) [inline core/modules/sink.h]: I was not sure, I will look through the `Queue*` things to check, but at least `QueueInc` and `QueueOut` are not thread safe (they read from PMD queue).

Comment (apanda) [inline core/modules/sink.h]: Last commit should have fixed the concern with queue. Will walk through all the modules in general.

File: core/module.cc [hunk 3]
```diff
@@ -265,11 +265,12 @@ placement_constraint Module::ComputePlacementConstraints() const {
 
 void Module::AddActiveWorker(int wid) {
   active_workers_.insert(wid);
-  LOG(WARNING) << "Added active worker " << wid << " to " << name_;
-  for (size_t i = 0; i < ogates_.size(); i++) {
-    if (ogates_[i]) {
-      auto next = static_cast<Module *>(ogates_[i]->arg());
-      next->AddActiveWorker(wid);
+  if (propagate_workers_) {
+    for (size_t i = 0; i < ogates_.size(); i++) {
```
Comment (barath) [inline core/module.cc]: Maybe use more idiomatic loop:
for (auto ogate : ogates_) ...?

**Comment by apanda** [comment]
@barath Resolved your comment.

**Comment by apanda** [comment]
@barath @sangjinhan @ddiproietto added the `attach_orphans` change from the e-mail.

**Comment by barath** [comment]
photo on 2012-05-12 at 13 54

**Pull request merged**